tsl/src/compression/compression.c

```diff
@@ -2539,65 +2666,158 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
+
+	foreach (lc, index_oids)
+	{
+		HeapTuple compressed_tuple;
+		Snapshot snapshot;
+		int num_segmentby_filtered_rows = 0;
+		int num_orderby_filtered_rows = 0;
 
-		switch (result)
+		Oid index_oid = lfirst_oid(lc);
+		Relation index_rel = index_open(index_oid, AccessShareLock);
+
+		snapshot = GetTransactionSnapshot();
+
+		IndexScanDesc scan =
+			index_beginscan(decompressor->in_rel, index_rel, snapshot, num_index_scankeys, 0);
+		TupleTableSlot *slot = table_slot_create(decompressor->in_rel, NULL);
+		index_rescan(scan, index_scankeys, num_index_scankeys, NULL, 0);
+		while (index_getnext_slot(scan, ForwardScanDirection, slot))
 		{
-			/* If the tuple has been already deleted, most likely somebody
-			 * decompressed the tuple already */
-			case TM_Deleted:
+			bool valid = false;
+			/* Deconstruct the tuple */
+			slot_getallattrs(slot);
+			compressed_tuple =
+				heap_form_tuple(slot->tts_tupleDescriptor, slot->tts_values, slot->tts_isnull);
+			compressed_tuple->t_self = slot->tts_tid;
+			num_segmentby_filtered_rows++;
+			if (num_scankeys)
 			{
-				if (IsolationUsesXactSnapshot())
+				/* filter tuple based on compress_orderby columns */
+				HeapKeyTest(compressed_tuple,
+							RelationGetDescr(decompressor->in_rel),
+							num_scankeys,
+							scankeys,
+							valid);
+				if (!valid)
 				{
-					/* For Repeatable Read isolation level report error */
-					table_endscan(heapScan);
-					ereport(ERROR,
-							(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
-							 errmsg("could not serialize access due to concurrent update")));
+					num_orderby_filtered_rows++;
+					continue;
 				}
-				continue;
 			}
-			break;
-			/*
-			 * If another transaction is updating the compressed data,
-			 * we have to abort the transaction to keep consistency.
-			 */
-			case TM_Updated:
-			{
-				table_endscan(heapScan);
-				elog(ERROR, "tuple concurrently updated");
-			}
-			break;
-			case TM_Invisible:
-			{
-				table_endscan(heapScan);
-				elog(ERROR, "attempted to lock invisible tuple");
-			}
-			break;
-			case TM_Ok:
-				break;
-			default:
+			heap_deform_tuple(compressed_tuple,
+							  decompressor->in_desc,
+							  decompressor->compressed_datums,
+							  decompressor->compressed_is_nulls);
+
+			if (!delete_compressed_tuple(decompressor, compressed_tuple))
 			{
-				table_endscan(heapScan);
-				elog(ERROR, "unexpected tuple operation result: %d", result);
+				heap_freetuple(compressed_tuple);
+				index_endscan(scan);
+				index_close(index_rel, AccessShareLock);
+				elog(ERROR, "unexpected tuple lock status");
 			}
-			break;
+
+			row_decompressor_decompress_row(decompressor, NULL);
+			heap_freetuple(compressed_tuple);
+			*chunk_status_changed = true;
 		}
-		row_decompressor_decompress_row(decompressor, NULL);
-		*chunk_status_changed = true;
+		elog(DEBUG1,
+			 "\nNumber of compressed rows fetched from index: %d \n"
+			 "Number of compressed rows filtered by orderby columns: %d",
```

Comment:
This message format does not conform to postgres guidelines about error messages. Have a look at https://www.postgresql.org/docs/current/error-style-guide.html

Reply:
Its not error message. Its debug message to know how many rows were scanned from index and how many rows were filtered.
Please suggest alternative if message is not appropriate.